### ISSUE TITLE
Incoordination between rewrite/cache plugins

### ISSUE THREAD

**Issue by jasonjoo2010** [miekg/coredns#6915]
Hi community,  

We deeply used rewrite plugin internally on CoreDNS for various purposes, like traffic redirection and application performance improvement, and it worked quite good and never disappointed users. But we encountered a strange issue and finally confirmed it's one thing we need to resolve in the code logic to make the rewrite/cache plugins better coordinated. After fixing it internally, I want to post the story here and ask about people's opinion if it's a necessary fix for open source.  

As people all know, the order of plugins getting called is very important in different aspects of performance and plugin effect. For example, the rate limit plugin should be in the front tier to block unwanted requests. The default call order for cache/rewrite is cache first, rewrite next. Meaning CoreDNS will use more space(memory) to trade for time(CPU). We swap the order of these 2 plugins considering the appropriate amount of rewrite rules doesn't consume much CPU, meanwhile we want to reduce the cache cost in some cases:  

An internal client library uses a UUID style domain when accessing the storage backend by design. The domain is in this form: `<uuid prefix>.store.intl.x-corp.com`. This behaviour exhausts the memory cache shortly and CoreDNS begins evicting cache entries. Although the cache eviction is not a big problem and it still runs efficiently, the queries are still not cached by CoreDNS. Furthermore, we usually take the evicting rate as one of the performance relevant indicators for config tuning. So we added new rewrite rule for this UUID domain to remove the UUID part from it to accelerate the resolution. But without moving the cache plugin behind the rewrite plugin, it won't be cached as expected. So we adjusted their order for this purpose and deployed to all production.  

All things are good until we encountered a strange issue: the domain name in the answer sometimes doesn't match the name in the query. It indicates the name reversion is not working although the rewrite rule works under certain conditions. After inspecting this issue, we get the reason as follows.  

- There is a rewrite rule of suffix type rewriting `.com.internal.domain` to `.com`.
- Client A queries `TEST1.EXAMPLE.COM` and get an answer for `TEST1.EXAMPLE.COM` which is good.
- Client B queries `test1.example.com` and get an answer for `TEST1.EXAMPLE.COM` which is a bit strange but still good. Because the domain resolution is case insensitive in the design.
- Client C queries `test1.example.com.internal.domain` and get an answer for `TEST1.EXAMPLE.COM` which is not good. The name in the answer doesn't match the name in the query so the client won't accept it.

There are basically 2-3 ways to fix it:  

- Persist the domain name in the cache in lower case only.
- Regard the case in the query.
- (Simplest) Revert the answer in case insensitive manner.

Discussion is welcomed.

Regards,

**What happened**:  

The domain resolution should be case insensitive but in the rewrite plugin when it reverts the name in the answer, it's case sensitive.  

eg.
Assuming `TEST1.EXAMPLE.COM` has been cached already.  
Assuming there is a suffix rewrite rule rewriting `.com.internal.domain` to `.com`.  
Query `test1.example.com.internal.domain` but get `TEST1.EXAMPLE.COM` in the answer.  

**What you expected to happen**:

The name in the answer should match the name in the query.

**How to reproduce it (as minimally and precisely as possible)**:

- Adjust the rewrite plugin in front of the cache plugin.
- There is a rewrite rule of suffix type rewriting `.com.internal.domain` to `.com`.
- Client A queries `TEST1.EXAMPLE.COM` and get an answer for `TEST1.EXAMPLE.COM` which is good.
- Client B queries `test1.example.com` and get an answer for `TEST1.EXAMPLE.COM` which is a bit strange but still good. Because the domain resolution is case insensitive in the design.
- Client C queries `test1.example.com.internal.domain` and get an answer for `TEST1.EXAMPLE.COM` which is not good. The name in the answer doesn't match the name in the query so the client won't accept it.

**Anything else we need to know?**:

**Environment**:

- the version of CoreDNS:
- Corefile:
- logs, if applicable:
- OS (e.g: `cat /etc/os-release`):
- Others:

**Comment by chrisohaver** [comment]
> Adjust the rewrite plugin in front of the cache plugin.

This first step kind of takes this out of "bug" territory, since it involves re-ordering the plugins and re-compiling.

We need to be extra careful not introduce any regression in fixing this, since it only affects folks that have custom compiled CoreDNS with the rewrite and cache plugin order swapped.

**Comment by jasonjoo2010** [comment]
> > Adjust the rewrite plugin in front of the cache plugin.
> 
> This first step kind of takes this out of "bug" territory, since it involves re-ordering the plugins and re-compiling.
> 
> We need to be extra careful not introduce any regression in fixing this, since it only affects folks that have custom compiled CoreDNS with the rewrite and cache plugin order swapped.

I agree what you emphasised here and we need to keep it safe. It may be not a real fix but makes it better implement the standard protocol. Different implementations may have difference in replies. Eg. Microsoft DNS is stricter in regarding the case of domain in the queries.  

Coming to the order of plugins, their call order was made configurable which is great. It gives users more flexibility to tweak/adjust on their own purposes when it makes sense. Although people may make mistakes under misunderstandings, they are able to correct after learning more from that.

Coming to this use case, the purpose/motivation behind swapping rewrite/cache plugins is as follows:  

- Reduce the cached entries.
- Reduce the unnecessary cache updates on the same domain after rewritten. It not only eases the stress of upstream DNS especially with prefetch enabled but also improves the resolution consistency. Because different domains which are rewritten to the same one share the same cache value.

After swapping the order, it worked as expected nicely until several months later, some users came and reported some strange cases. It's difficult to reproduce because it only happens when the queries from clients in different systems reach to DNS in certain orders. From user perspective, the rewrite plugin fails to revert the domain name back in the response while the response is marked as a success. Which looks more like a surprise.
